src/Database.cpp

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright (c) 2021 Chiu Yen-Chen, Swen Sun-Yen, Wen Yong-Wei, Yuan Wei-Chen.
+All rights reserved.
+Use of this source code is governed by a BSD-style license that can be
+found in the LICENSE file. See the AUTHORS file for names of contributors.
+ */
+
+#include "Database.h"
+
+namespace shortlink {
+std::string Database::CreateSURL(const std::string &raw_url) {
+  // Create the short url
+  std::string short_url = SURLCreator::CreateSURL(raw_url);
+
+  // Check if short url exists
+  std::string read_url = ReadSURL(short_url);
+  if (read_url != "") {
+    // if short url exists
+    if (read_url == raw_url) {
+      // the short url and its corresponding raw_url in lut is the same as the
+      // input raw_url. Update the surl expire time.
+      Date now = boost::gregorian::day_clock::local_day();
+      UpdateSURL(std::move(lut.at(short_url)), now);
+    } else {
+      // the short url in lut is the same as the input short url,
+      // but the corresponding raw_url is not the same.
+      // Add the offset to Encode the url
+      int offset = 1;
+      while (read_url != raw_url) {
+        short_url = SURLCreator::CreateSURL(raw_url, offset);
+        read_url = ReadSURL(short_url);
+      }
+      std::unique_ptr<SURL> surl(new SURL(raw_url, short_url));
+      InsertSURL(short_url, std::move(surl));
+    }
+  } else {
+    // if short url does not in the lut, insert it to the lut
+    std::unique_ptr<SURL> surl(new SURL(raw_url, short_url));
+    InsertSURL(short_url, std::move(surl));
+    return short_url;
+  }
+}
+std::string Database::ReadSURL(const std::string &surl) const {
+  // if the surl is not in the lut, return "", else return the raw url.
```

Comment:
Consider locking `write_lock` here as well, unless you are sure that the lut update is atomic. i.e. If we read while we're writing, could it cause any error or incorrect read?